### PR TITLE
New extension for ActivityResultManager

### PR DESCRIPTION
**Pull request by TatianaKamaletdinova**

### CODE REVIEW

**Comment by hichamboushaba** [comment]
Thanks @TatyanaBazhanova for the PR, and sorry for the delay, I was just a bit busy with some stuff, I'll try to take a look this week.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt [hunk 1]
```diff
@@ -21,7 +21,7 @@ internal object ActivityProvider : Application.ActivityLifecycleCallbacks {
             val activity = it.get()
             activity != null && activity.lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED)
         }
-        .map { it.get()!! }
+        .map { it.get() }
```
Comment (hichamboushaba) [inline library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: Please revert this, this breaks the code on other parts, since it changes the type of `activityFlow` from `Flow<ComponentActivity>` to `Flow<ComponentActivity?>`

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt [hunk 2]
```diff
@@ -56,5 +56,7 @@ internal object ActivityProvider : Application.ActivityLifecycleCallbacks {
 
     override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {}
 
-    override fun onActivityDestroyed(activity: Activity) {}
+    override fun onActivityDestroyed(activity: Activity) {
+        _activityFlow.value = WeakReference(null)
+    }
```
Comment (hichamboushaba) [inline library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: Please revert this, this would might break the logic on multi-activities apps, and this is not needed because:
1. The `WeakReference` will prevent leakage.
2. We check if the activity is still active [here](https://github.com/hichamboushaba/SuspendActivityResult/blob/05a26fe0b711f5f669e2c14eb92b8e670c228e13/library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt#L22) before returning it.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt [hunk 3]
```diff
@@ -90,4 +92,18 @@ suspend fun ActivityResultManager.openDocumentTree(startingLocation: Uri? = null
...
 suspend fun ActivityResultManager.createDocument(fileName: String): Uri? {
     return requestResult(CreateDocument(), fileName)
-}
\ No newline at end of file
+}
+
+/**
+ * see [StartIntentSenderForResult]
+ */
+suspend fun ActivityResultManager.takeActivityResult(intentSenderRequest: IntentSenderRequest): ActivityResult? {
```
Comment (hichamboushaba) [inline library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt]: Let's rename this to the same name that the default function uses: [`startIntentSender`](https://developer.android.com/reference/android/app/Activity.html#startIntentSender-android.content.IntentSender-android.content.Intent?-kotlin.Int-kotlin.Int-kotlin.Int-)

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt [hunk 4]
```diff
@@ -90,4 +92,18 @@ suspend fun ActivityResultManager.openDocumentTree(startingLocation: Uri? = null
...
+    return requestResult(StartIntentSenderForResult(), intentSenderRequest)
+}
+
+/**
+ * see [StartActivityForResult]
+ */
+suspend fun ActivityResultManager.takeStartActivityResult(intent: Intent): ActivityResult? {
```
Comment (hichamboushaba) [inline library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt]: I think a name like `startActivityForResult` would be clearer on what the function does.

**Review by hichamboushaba — CHANGES_REQUESTED** [review]
Thanks again @TatyanaBazhanova for the PR, I left some comments that need to be fixed before merging this.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt (hunk 2 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: Thank you for comment. I didn't understand it before, because i have single activity in my project. I reverted it.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt (hunk 1 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: I think then better to add .filterNotNull() to _activityFlow, because if it.get() return null it will be NPE. It can be bad and dangerous. My variant is that:

```kotlin
val activityFlow: Flow<ComponentActivity> =
        _activityFlow
            .asStateFlow()
            .map { it.get() }
            .filterNotNull()
            .filter { it.lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED) }
```

Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: And applying 'distinctUntilChanged' to StateFlow has no effect. Because StateFlow already behaves as if distinctUtilChanged operator is applied

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt (hunk 3 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt]: fix

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt (hunk 4 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/Helpers.kt]: fix

**Comment by TatianaKamaletdinova** [comment]
Thanks @hichamboushaba for your comments, i fixed some comments and answered on comment about activityFlow.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt (hunk 1 above)
Comment (hichamboushaba) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: > And applying 'distinctUntilChanged' to StateFlow has no effect. Because StateFlow already behaves as if distinctUtilChanged operator is applied

`StateFlow` will use the default `equals` implementation for comparison, that's not what we are doing, we are applying the `distinctUtilChanged` on the `Activity` instance, since we may send the same activity multiple times, in different `WeakReference` instances.

> I think then better to add .filterNotNull() to _activityFlow, because if it.get() return null it will be NPE. It can be bad and dangerous. My variant is that:

This is the same logic as I'm doing, I'm already filtering non-null activity instances on the `filter` block, so `it.get()` won't throw any NPE.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityResultManagerImpl.kt [hunk 5]
```diff
@@ -48,7 +48,7 @@ internal object ActivityResultManagerImpl : ActivityResultManager {
                         }
 
                         if (!isLaunched) {
-                            launcher!!.launch(input)
+                            launcher?.launch(input)
```
Comment (hichamboushaba) [inline library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityResultManagerImpl.kt]: I think I wrote a comment about this, but for some reason it isn't not there 😕 

Please revert this change, I prefer a crash instead of having a non-expected state ([offensive programming](https://en.wikipedia.org/wiki/Offensive_programming#:~:text=Offensive%20programming%20is%20a%20name,are%20not%20fundamentally%20in%20conflict.))

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt (hunk 1 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: > StateFlow will use the default equals implementation for comparison, that's not what we are doing, we are applying the distinctUtilChanged on the Activity instance, since we may send the same activity multiple times, in different WeakReference instances.

It's important information, then could be to add comment for it case. I fixed:
https://github.com/hichamboushaba/SuspendActivityResult/pull/1/commits/60991fde23e1876eb11a56ce128e71c04b14b9b0

Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: > This is the same logic as I'm doing, I'm already filtering non-null activity instances on the filter block, so it.get() won't throw any NPE.

I do not agree with you.
1. WeakReference can be null anytime and it can be happen in moment then will executing .map { it.get()!! } and then will be NPE. 
2. In first initialization there is a chance can be NPE, because MutableStateFlow emit null: 
> private val _activityFlow = MutableStateFlow(WeakReference<ComponentActivity>(null))

And finally, why ide highlights in place for insert "!!", because kotlin is null safety and do it that it do kotlin contract and filtering non-null activity instances on the filter block is not checking on kotlin-contract. It do .filterNotNull().

If that reasons don't look convincing to you, i will fix it, but i will be sad.

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityResultManagerImpl.kt (hunk 5 above)
Comment (TatianaKamaletdinova) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityResultManagerImpl.kt]: https://github.com/hichamboushaba/SuspendActivityResult/pull/1/commits/36e8f06d90bc71db8a9d529391b075df6e430a84

File: library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt (hunk 1 above)
Comment (hichamboushaba) [reply library/src/main/java/dev/hichamboushaba/suspendactivityresult/ActivityProvider.kt]: > In first initialization there is a chance can be NPE, because MutableStateFlow emit null:

As I said before, it won't cause any NPE, I had a check on null values on the `filter` block before 🙂.

I'm merging this as is, since it's the same logic as before, and won't hurt.
But just for future contributions, please limit changes to the relevant parts of the PR 🙏 , for example for this, it should've been just the new extensions, if you want to change something else, we can discuss it in a separate PR 🙂.

**Review by hichamboushaba — APPROVED** [review]
Thanks again @TatyanaBazhanova for the contribution.

**Comment by TatianaKamaletdinova** [comment]
Thanks you for best library @hichamboushaba sorry and thanks for discussion 👍🏼  (p.s. it was my first contribution, next time i'll take your recommendations)

**Comment by hichamboushaba** [comment]
Congrats on your first contribution @TatyanaBazhanova, and really good work.
